### PR TITLE
8283307: Vectorize unsigned shift right on signed subword types

### PR DESCRIPTION
**Pull request by fg1417**
```
public short[] vectorUnsignedShiftRight(short[] shorts) {
    short[] res = new short[SIZE];
    for (int i = 0; i < SIZE; i++) {
        res[i] = (short) (shorts[i] >>> 3);
    }
    return res;
}
```
In C2's SLP, vectorization of unsigned shift right on signed subword types (byte/short) like the case above is intentionally disabled[1]. Because the vector unsigned shift on signed subword types behaves differently from the Java spec. It's worthy to vectorize more cases in quite low cost. Also, unsigned shift right on signed subword is not uncommon and we may find similar cases in Lucene benchmark[2].

Taking unsigned right shift on short type as an example,
image

when the shift amount is a constant not greater than the number of sign extended bits, 16 higher bits for short type shown like
above, the unsigned shift on signed subword types can be transformed into a signed shift and hence becomes vectorizable. Here is the transformation:
image

This patch does the transformation in `SuperWord::implemented()` and `SuperWord::output()`. It helps vectorize the short cases above. We can handle unsigned right shift on byte type in a similar way. The generated assembly code for one iteration on aarch64 is like:
```
...
sbfiz   x13, x10, #1, #32
add     x15, x11, x13
ldr     q16, [x15, #16]
sshr    v16.8h, v16.8h, #3
add     x13, x17, x13
str     q16, [x13, #16]
...
```

Here is the performance data for micro-benchmark before and after this patch on both AArch64 and x64 machines. We can observe about ~80% improvement with this patch.

The perf data on AArch64:
Before the patch:
Benchmark        (SIZE)  (shiftCount)  Mode  Cnt    Score   Error  Units
urShiftImmByte    1024         3       avgt    5  295.711 ± 0.117  ns/op
urShiftImmShort   1024         3       avgt    5  284.559 ± 0.148  ns/op

after the patch:
Benchmark         (SIZE) (shiftCount)  Mode  Cnt    Score   Error  Units
urShiftImmByte     1024        3       avgt    5   45.111 ± 0.047  ns/op
urShiftImmShort    1024        3       avgt    5   55.294 ± 0.072  ns/op

The perf data on X86:
Before the patch:
Benchmark        (SIZE) (shiftCount)  Mode  Cnt    Score    Error  Units
urShiftImmByte    1024        3       avgt    5  361.374 ±  4.621  ns/op
urShiftImmShort   1024        3       avgt    5  365.390 ±  3.595  ns/op

After the patch:
Benchmark        (SIZE) (shiftCount)  Mode  Cnt    Score    Error  Units
urShiftImmByte    1024        3       avgt    5  105.489 ±  0.488  ns/op
urShiftImmShort   1024        3       avgt    5   43.400 ±  0.394  ns/op

[1] https://github.com/openjdk/jdk/blob/002e3667443d94e2303c875daf72cf1ccbbb0099/src/hotspot/share/opto/vectornode.cpp#L190
[2] https://github.com/jpountz/decode-128-ints-benchmark/

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8283307](https://bugs.openjdk.java.net/browse/JDK-8283307): Vectorize unsigned shift right on signed subword types

### Reviewers
 * [Jie Fu](https://openjdk.java.net/census#jiefu) (@DamonFool - **Reviewer**)
 * [Pengfei Li](https://openjdk.java.net/census#pli) (@pfustc - Committer)
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7979/head:pull/7979` \
`$ git checkout pull/7979`

Update a local copy of the PR: \
`$ git checkout pull/7979` \
`$ git pull https://git.openjdk.java.net/jdk pull/7979/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7979`

View PR using the GUI difftool: \
`$ git pr show -t 7979`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7979.diff">https://git.openjdk.java.net/jdk/pull/7979.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/superword.cpp
Comment (DamonFool) [inline src/hotspot/share/opto/superword.cpp:2029]: I'm still not clear why `>>>` can be transferred to `>>` iff `shift_cnt <= the number of sign extended bits`.

File: src/hotspot/share/opto/superword.cpp
Comment (fg1417) [reply src/hotspot/share/opto/superword.cpp:2027]: In any Java arithmetic operation, operands of small integer types (boolean, byte, char & short) should be promoted to int first. For example, for negative short value, after sign-extension to int, the value should be like:
image
In java spec, unsigned right shift on the promoted value is to shift data right and fill the higher bits with zero-extension. We may find that when shift amount is less than 16, the lower-16 bit value is right shift with one-extension, like:
image
As vector elements of small types don't have upper bits of int, vector unsigned right shift on short elements is to fill lower bits with 0 directly like:
image
In this way, the result of vector unsigned right shift is different from the result of scalar unsigned right shift for signed subword types.

File: src/hotspot/share/opto/superword.cpp
Comment (fg1417) [reply src/hotspot/share/opto/superword.cpp:2029]: When shift_cnt <= the number of sign extended bits, vector signed right shift can work the same as scalar unsigned right shift for subword types in the case I mentioned above, when short value is negative. Vector right shift can fill the shifted bits with sign-extension as we expected. As for positive short values, n >>> s  works the same as n >> s in Java spec.

File: src/hotspot/share/opto/superword.cpp
Comment (DamonFool) [reply src/hotspot/share/opto/superword.cpp:2027]: > In any Java arithmetic operation, operands of small integer types (boolean, byte, char & short) should be promoted to int first. For example, for negative short value, after sign-extension to int, the value should be like: image In java spec, unsigned right shift on the promoted value is to shift data right and fill the higher bits with zero-extension. We may find that when shift amount is less than 16, the lower-16 bit value is right shift with one-extension, like: image As vector elements of small types don't have upper bits of int, vector unsigned right shift on short elements is to fill lower bits with 0 directly like: image In this way, the result of vector unsigned right shift is different from the result of scalar unsigned right shift for signed subword types.

Got it.
Thanks for your kind explanation.

**Comment by fg1417** [comment]
> However, changing the opcode in superword code seems tricky, which may be not easy to maintain. Why not transform scalar `urhsift` --> `rshift` during the GVN phase like this?

Thanks for your advice.

The simple transformation from `urshift` to `rshift` is based on vector operation and limited shift amount as described in comment lines. We can't easily do the replacement from `urshift` to `rshift` in GVN phase. That's because scalar `rshift` on subword types are not equivalent to scalar `urshift` on subword types in Java Spec. In GVN phase, we still have no idea if the current node can be vectorized and have to follow the quite complex rules if we do the transformation, as illustrated here https://docs.oracle.com/javase/specs/jls/se18/html/jls-15.html#jls-15.19.  It would bring extra overhead for all urshift operations on subword types whether they can be vectorized or not.

Besides, when we do the transformation in SLP, the urshift operations on subword types are not always followed by `Store` node and they can be intermediate results. In that case, we can still do the transformation and help vectorize them in SLP phase. But we will miss the chance if in GVN phase.

In conclusion, code change in SLP phase seems much easier and can cover more potential scenarios.

**Comment by DamonFool** [comment]
Okay, you're right.
It seems impossible to replace all the target `urshift` with `rshift` during GVN phase.

Instead of just changing the opcode, can we replace `urshift` with `rshift` during SLP phase?

**Comment by fg1417** [comment]
> Instead of just changing the opcode, can we replace `urshift` with `rshift` during SLP phase?

Sorry, what does "replace `urshift` with `rshift`" mean? Could you please explain it in detail? Thanks @DamonFool .

 Actually, we can only replace `urshift` with `rshift` in the stage of generating vector node, https://github.com/openjdk/jdk/blob/c35590282d54d8388f2f7501a30365e0a912bfda/src/hotspot/share/opto/superword.cpp#L2384, to guarantee its correctness. SLP may breaks off in any stage before it and, if so, we can't do the replacement because of the same reason as in GVN phase.

**Comment by DamonFool** [comment]
> > Instead of just changing the opcode, can we replace `urshift` with `rshift` during SLP phase?
> 
> Sorry, what does "replace `urshift` with `rshift`" mean? Could you please explain it in detail? Thanks @DamonFool .
> 
> Actually, we can only replace `urshift` with `rshift` in the stage of generating vector node,
> 
> https://github.com/openjdk/jdk/blob/c35590282d54d8388f2f7501a30365e0a912bfda/src/hotspot/share/opto/superword.cpp#L2384
> 
> , to guarantee its correctness. SLP may breaks off in any stage before it and, if so, we can't do the replacement because of the same reason as in GVN phase.

Once `SuperWord::compute_vector_element_type()` computes the type of `urshift` is `short/byte`, then it can be transformed to `rshift` iff shift_cnt <= {16/24}.
In that case, it would be still correct even though the SLP analysis may fail, right?

Maybe, we can implement this idea like this: https://github.com/openjdk/jdk/pull/8224 .
What do you think?

**Comment by fg1417** [comment]
> Once `SuperWord::compute_vector_element_type()` computes the type of `urshift` is `short/byte`, then it can be transformed to `rshift` iff shift_cnt <= {16/24}. In that case, it would be still correct even though the SLP analysis may fail, right?
> 
> Maybe, we can implement this idea like this: #8224 . What do you think?

Thanks a lot for your explanation and your code, @DamonFool 

I suppose this idea https://github.com/openjdk/jdk/pull/8224 works and would be still correct even though the SLP analysis fails.

But we don't have to do the replacement when vectorization fails, right? And, if we want to understand the new implementation in your code, we still need to know the background and the relationship between `rshift` and `urshift` on subword types.  The idea in this pr is intended to put off the replacement to the necessary stage. Also, changing the vector opcode is a simple and low-cost way to do the replacement. WDYT?

**Comment by DamonFool** [comment]
> I suppose this idea #8224 works and would be still correct even though the SLP analysis fails.
> 
> But we don't have to do the replacement when vectorization fails, right? And, if we want to understand the new implementation in your code, we still need to know the background and the relationship between `rshift` and `urshift` on subword types. The idea in this pr is intended to put off the replacement to the necessary stage. Also, changing the vector opcode is a simple and low-cost way to do the replacement. WDYT?

I agree with you.
After hours of work to implement https://github.com/openjdk/jdk/pull/8224 last night, I think your patch is excellent and smart enough.

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (DamonFool) [inline test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: This `requires` would disable the test on some x86 machines.
So we'd better fix it.

**Comment by DamonFool** [comment]
Please also update the comments in the following tests.
```
compiler/vectorization/runner/ArrayShiftOpTest.java
compiler/vectorization/runner/BasicByteOpTest.java
compiler/vectorization/runner/BasicShortOpTest.java
```

E.g., remove comments like this
```
    @Test
    // Note that unsigned shift right on subword signed integer types can't
    // be vectorized since the sign extension bits would be lost.
    public short[] vectorUnsignedShiftRight() {
        short[] res = new short[SIZE];
        for (int i = 0; i < SIZE; i++) {
            res[i] = (short) (shorts2[i] >>> 3);
        }
        return res;
    }
```

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (DamonFool) [inline test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:114]: This test won't work as expected if `(byte) (bytea[i] >>> 3)` is also vectorized.
So another question: can we make sure `(byte) (bytea[i] >>> 3)` wouldn't be vectorized during testing?

Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:114]: Yes, the argument `(byte) (bytea[i] >>> 3)` of the function call won't be vectorized. Even if the function call is inlined, https://github.com/openjdk/jdk/blob/339005dbc99e94ed094612c7b34eb0c93ca1f8c1/test/lib/jdk/test/lib/Asserts.java#L200, any control flow statements can't be vectorized by SLP https://github.com/openjdk/jdk/blob/339005dbc99e94ed094612c7b34eb0c93ca1f8c1/src/hotspot/share/opto/superword.cpp#L125. Besides I suppose the function call is hard to be inlined.

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:114]: > Besides I suppose the function call is hard to be inlined.

Then I would suggest running with `-XX:-Inline`.

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: I notice that the file checked some vector nodes like `LOAD_VECTOR` but there is no requirement on machine https://github.com/openjdk/jdk/blob/d41331e6f2255aa07dbbbbccf62e39c50269e269/test/hotspot/jtreg/compiler/c2/irTests/TestAutoVectorization2DArray.java#L32. May I ask why? I suppose that not all machines support simd and the check for vector node may fail, right? I can't enable GHA because of some unknown limitation on my account. I fear that fixing it here may break GHA. Thanks.

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: > I notice that the file checks some vector nodes like `LOAD_VECTOR` but there is no requirement on machine
> 
> https://github.com/openjdk/jdk/blob/d41331e6f2255aa07dbbbbccf62e39c50269e269/test/hotspot/jtreg/compiler/c2/irTests/TestAutoVectorization2DArray.java#L32
> 
> . May I ask why? I suppose that not all machines support simd and the check for vector node may fail, right? I can't enable GHA because of some unknown limitation on my account. I fear that fixing it here may break GHA. Thanks.

LoadV/StoreV/AddV are basic and common operations which are supported by modern CPUs.
The test should be run on both x86 and aarch64 if you are not sure whether other CPUs support RShiftV.

**Comment by merykitty** [comment]
Hi,

My humble opinion is that we should not bother to optimise this pattern. An unsigned shift should operate on an unsigned type, so doing `(short)(s[i] >>> k)` is more likely a bug rather than an intended behaviour. This operation is really hard to reason about, and often not the result anyone cares about. The first point is that this operation cannot be reason as a simple shift, and must be viewed as an int shift between 2 casts. And the second point is that if you want an unsigned shift, you should cast the values to `int` in an unsigned manner.

Thank you very much.

**Comment by merykitty** [comment]
@fg1417 Thanks for your response, I have taken a look at the benchmark, it seems that the full operation is `((short & 0xFFFF) >>> k) & (1 << l - 1)` and then the `& 0xFFFF` part is omitted if `k + l <= 16`. So it would be the most appropriate if we can do the same here, that is to recognise the pattern `(short & x) >>> k` and `(short >>> k) & y` with `x \in 0xFFFF` and `y \in (1 << (16 - k) - 1)` (`\in` as in every set bit of the first operand is also a set bit of the second operand).

Thank you very much.

**Comment by fg1417** [comment]
> @fg1417 Thanks for your response, I have taken a look at the benchmark, it seems that the full operation is `((short & 0xFFFF) >>> k) & (1 << l - 1)` and then the `& 0xFFFF` part is omitted if `k + l <= 16`. So it would be the most appropriate if we can do the same here, that is to recognise the pattern `(short & x) >>> k` and `(short >>> k) & y` with `x \in 0xFFFF` and `y \in (1 << (16 - k) - 1)` (`\in` as in every set bit of the first operand is also a set bit of the second operand).

Thanks for your kind reply, @merykitty .

Actually, the pattern `(short & x) >>> k` has been transformed into `(short  >>> k) & (x >>> k)` in GVN phase https://github.com/openjdk/jdk/blob/21ea740e1da48054ee46efda493d0812a35d786e/src/hotspot/share/opto/mulnode.cpp#L1305. In this way, SLP need to support only `(short  >>> k)` like the patch did, to help vectorize the whole `(short & x) >>> k` or `(short >>> k) & y`, because the rest part,  ` & (x >>> k)`, has been supported by SLP already. 

If the optimization in GVN phase doesn't work and the pattern `(short & x) >>> k` is transferred to SLP phase, SLP won't take  `(short & x) >>> k` as potential short vector operations. Because C2 can't get precise info about sign here for the shift operation, https://github.com/openjdk/jdk/blob/21ea740e1da48054ee46efda493d0812a35d786e/src/hotspot/share/opto/superword.cpp#L3328. If the src input of the `URShift` is not from a `load` node, we can't assign it as any signed subword type and vectorization will break off. That's to keep strictly consistent with Java Spec in `Shift` behavior.

Thank you very much.

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java]: > This `requires` would disable the test on some x86 machines. So we'd better fix it.

Done. Thanks.

**Comment by fg1417** [comment]
> Please also update the comments in the following tests.

Done. Thanks for your review. @DamonFool

**Comment by merykitty** [comment]
@fg1417 Thanks a lot for your kind explanation, it makes sense now. Do you think it is worth it to generalise to the remaining cases, that is

    0 < c < esize:
    (type)((int)a >>> c) -> a >> c

    esize <= c <= 32 - esize:
    (type)((int)a >>> c) -> a >> (esize - 1)

    32 - esize < c < 32:
    (type)((int)a >>> c) -> (a >> (esize - 1)) >>> (c - (32 - esize))

Here `>>` and `>>>` are true unsigned shifts on the operand types, not exactly the Java scalar operations which actually work on promoted types.

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (DamonFool) [inline test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:114]: I'm still a bit worried about the test.

Suggestion:
Rewrite
```
Asserts.assertEquals(byteb[i], (byte) (bytea[i] >>> 3));
```

to
```
Asserts.assertEquals(byteb[i], urshift(bytea[i], 3)));
```
And disable inlining during the testing.

**Comment by fg1417** [comment]
> ```
> 0 < c < esize:
> (type)((int)a >>> c) -> a >> c
> 
> esize <= c <= 32 - esize:
> (type)((int)a >>> c) -> a >> (esize - 1)
> 
> 32 - esize < c < 32:
> (type)((int)a >>> c) -> (a >> (esize - 1)) >>> (c - (32 - esize))
> ```
> 
> Here `>>` and `>>>` are true unsigned shifts on the operand types, not exactly the Java scalar operations which actually work on promoted types.

Thanks for your kind reply, @merykitty . 

I suppose the first two scenarios have been covered in the pr. `esize` for byte is 8 and `esize` for short is 16, and the pr covers the range from 0-24 for byte and the range from 0-16 for short. But may I ask why the shift amount is `esize-1` rather than `esize` itself in the second scenario when `esize <= c <= 32 - esize`? 

For the third scenario, the idea works and generalizing to cover more cases is absolutely good. However, the true unsigned shifts `>>>` on the right of induction may make people confused, if there are two different unsigned right shift vector operations on the same data type in one patch. It's not easy to review. I mean maybe we can do it with another patch. WDYT? 

Thanks a lot.

**Comment by merykitty** [comment]
@fg1417

> But may I ask why the shift amount is `esize-1` rather than `esize` itself in the second scenario when `esize <= c <= 32 - esize`?

Because `a >> esize == a`, similar to how scalar integer and long shift works, vector shift on bytes and shorts mask the shift count to the least significant 3, 4 bits, respectively.

> However, the true unsigned shifts `>>>` on the right of induction may make people confused, if there are two different unsigned right shift vector operations on the same data type in one patch.

It is not that confusing I think, vector shifts always consider bytes and shorts as first-class types, not dress-up ints as in scalar circumstances, but it is upon your decision.

Thanks.

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java]: Done. Thanks very much @DamonFool .

**Review by DamonFool — APPROVED** [review]
LGTM

Thanks for the update.

**Comment by fg1417** [comment]
Thanks for your review, @DamonFool . Can I get a second review please?

**Comment by fg1417** [comment]
Can I get a second review please :)

**Comment by fg1417** [comment]
Hi @vnkozlov @rwestrel , can I get a second review please :)

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (pfustc) [inline test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: This IR test for vectorizable check looks good on AArch64. But AFAIK, some operations cannot be vectorized on old x86 CPUs with AVX=1. Could you add something like `(os.simpleArch == "x64" & vm.cpu.features ~= ".*avx2.*")` to check the CPU feature?

**Review by pfustc — COMMENTED** [review]
LGTM except my concern in the IR test.

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (pfustc) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: @merykitty @sviswa7 Could you help confirm if byte/short shift operations are vectorizable with all AVX versions of x86?

**Review by pfustc — APPROVED** [review]

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (sviswa7) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: @pfustc They are available either directly or through a series of instructions for majority of the architectures. All implemented in x86.ad. The match_rule_supported and match_rule_supported_vector will appropriately return true/false.

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: RShiftVB would fail with `UseSSE < 4`.
But is there a x86 machine running jtreg tests with `UseSSE < 4`?

For `RShiftV{B,S}`, neither `match_rule_supported ` nor `match_rule_supported_vector ` would return false on x86 with AVX=1.
So the test seems fine on x86, right? @sviswa7 @pfustc

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: > But is there a x86 machine running jtreg tests with `UseSSE < 4`?
> 
To clarify: is there a x86 cpu which only supports `UseeSSE < 4` ISA running jtreg tests?

Comment (vnkozlov) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: We do run vectorization (superword, vectorization, VectorAPI) tests in our testing with:
`-XX:UseAVX=0 -XX:UseSSE=3` and `-XX:UseAVX=0 -XX:UseSSE=2`

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: > We do run vectorization (superword, vectorization, VectorAPI) tests in our testing with: `-XX:UseAVX=0 -XX:UseSSE=3` and `-XX:UseAVX=0 -XX:UseSSE=2`


`compiler/c2/irTests/TestVectorizeURShiftSubword.java` still passed with `-XX:UseAVX=0 -XX:UseSSE=3` and `-XX:UseAVX=0 -XX:UseSSE=2`.

Maybe the IR test framework wouldn't accept VM args like `-XX:UseAVX=? -XX:UseSSE=?`?

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: Two questions @fg1417  :
1. Is `RShiftVB` generated with `-XX:UseAVX=0 -XX:UseSSE=2` on x86?
2. If the answer is no for question 1, why `TestVectorizeURShiftSubword.java` get passed?
Thanks.

Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: > Two questions @fg1417 :
> 
> 1. Is `RShiftVB` generated with `-XX:UseAVX=0 -XX:UseSSE=2` on x86?
> 2. If the answer is no for question 1, why `TestVectorizeURShiftSubword.java` get passed?
>    Thanks.

@DamonFool , no `RShiftB` nodes are generated with `-XX:UseAVX=0 -XX:UseSSE=2` on x86. The test passed because when adding VM options `-XX:UseAVX=0 -XX:UseSSE=2` externally, the IR framework won't do the IR check and just run the java code, I verified and made the guess. But I'm not clear whether all external VM options would stop the IR framework to do the IR check.

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: > 
> @DamonFool , no `RShiftB` nodes are generated with `-XX:UseAVX=0 -XX:UseSSE=2` on x86. The test passed because when adding VM options `-XX:UseAVX=0 -XX:UseSSE=2` externally, the IR framework won't do the IR check and just run the java code, I verified and made the guess. But I'm not clear whether all external VM options would stop the IR framework to do the IR check.

Thanks for the verification.
Then, I think the test is fine.

What do you think? @vnkozlov

Comment (vnkozlov) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: As @fg1417 pointed, IR framework will not do any IR verification testing (it does not run test) when some flags are passed to test and not whitelisted (`UseAVX` and `UseSSE` are not on list):
https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/TestFramework.java#L105

I hacked VM to set `UseSSE=2 UseAVX=0` and test failed as expected because it did not find `RShiftVB` as you said:
```
One or more @IR rules failed:
Failed IR Rules (2) of Methods (2)
```
With `UseSSE=4 UseAVX=0` combination test passed. Our testing systems don't have anything less than SSE4 CPUs. That is why it passed our testing. I assume Github actions don't have such old systems too. 

**I don't see these changes passed pre-submit GitHub testing. Make sure they passed.**

Currently there are changes in review #8999 which allow to specify CPU features as condition for testing method.

I suggest to leave test as it is (if pre-submit GitHub testing passed). If we hit any issues later we will use #8999 API to fix it.

**Review by sviswa7 — APPROVED** [review]
Very nice work! Patch looks good to me.

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: > I don't see these changes passed pre-submit GitHub testing. Make sure they passed.

I don't know why the pre-submit GitHub testing didn't run in this PR.
So I create a draft PR for this path to run the pre-submit testing: https://github.com/openjdk/jdk/pull/9026 .

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: > As @fg1417 pointed, IR framework will not do any IR verification testing (it does not run test) when some flags are passed to test and not whitelisted (`UseAVX` and `UseSSE` are not on list):
> https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/TestFramework.java#L105

Ah, I see.
Thanks @vnkozlov for your explanation.

Comment (DamonFool) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: > > I don't see these changes passed pre-submit GitHub testing. Make sure they passed.
> 
> I don't know why the pre-submit GitHub testing didn't run in this PR. So I create a draft PR for this patch to run the pre-submit testing: #9026 .

The pre-submit testing finished without regression. @vnkozlov @fg1417 
Thanks.

**Review by vnkozlov — APPROVED** [review]
Thank you for verifying with GHA testing.

**Comment by fg1417** [comment]
Thanks for all your kind review, @vnkozlov @pfustc @sviswa7 .

File: test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java
Comment (fg1417) [reply test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java:36]: Thanks for your help on GHA testing @DamonFool .

**Pull request closed without merging**
